Tweetinvi.Core/Core/TweetinviSettingsAccessor.cs

```diff
@@ -39,104 +39,135 @@ public interface ITweetinviSettingsAccessor
         RateLimitTrackerMode RateLimitTrackerMode { get; set; }
 
         /// <summary>
-        /// How much additional time to wait than should be strictly necessary for a new batch of Twitter rate limits
-        /// to be available. Required to account for timing discrepancies both within Twitter's servers and between
-        /// Twitter and us. 
+        /// Specify whether you want exceptions coming from Twitter are returning null instead of throwing
         /// </summary>
+        bool OnTwitterExceptionReturnNull { get; set; }
+
         int RateLimitWaitFudgeMs { get; set; }
     }
 
-    public class TweetinviSettingsAccessor : ITweetinviSettingsAccessor
+    public class TweetinviSettingsAccessor : ITweetinviSettingsAccessor, IAsyncContextPreparable
     {
-        private static ITweetinviSettings StaticTweetinviSettings { get; set; }
+        private static ITweetinviSettings StaticTweetinviSettings { get; }
 
-        public TweetinviSettingsAccessor()
+        private static readonly AsyncLocal<ITweetinviSettings> _executionContextTweetinviSettings;
```

Comment:
I assume the reason there is now thread local, async local and a static instance is to try and allow a user to set things lower down the stack and still have them propagate back up, e.g.:
```cs
await setCredentialsAsync();
await updateUserAsync();
```
as if this was just using async local, `setCredentialsAsync` changes would be lost unless the async local value had already been set before calling the method.

However, it seems very complex and i'm wary of it hiding other problems. If we're going to do it this way, it needs to be tested to destruction.
I'd be tempted to change this to be purely async local and document how to go about using it. I can't think of a reason for users to need to update their settings asynchronously, so if we change the above example to:
```cs
setCredentials();
await updateUserAsync();
```
I think everything should work fine using async local.

What do you thinnk?

Reply:
> I assume the reason there is now thread local, async local and a static instance is to try and allow a user to set things lower down the stack and still have them propagate back up, e.g.:

Not really. As we have removed `ExceptionHandler` being Thread/Async Context specific, we no longer need to propagate back up.

What this is doing is ensuring that both Thread context and Async context behave the same way.

> However, it seems very complex and i'm wary of it hiding other problems. If we're going to do it this way, it needs to be tested to destruction.

I totally agree, hence the solution we discussed about the `TwitterClient`.
There are also limitations with what this also give us. For example it becomes very complex to handle ThreadContext within an AsyncContext and the other way around.

> I'd be tempted to change this to be purely async local and document how to go about using it. I can't think of a reason for users to need to update their settings asynchronously, so if we change the above example to:

This could make sense, but I thought it would be confusing for users to have 2 different approach to a same problem as we would have Thread created by async/await not behaving the same way as Threads created by the user.

Reply:
> > I assume the reason there is now thread local, async local and a static instance is to try and allow a user to set things lower down the stack and still have them propagate back up, e.g.:
> 
> Not really. As we have removed `ExceptionHandler` being Thread/Async Context specific, we no longer need to propagate back up.
> 
> What this is doing is ensuring that both Thread context and Async context behave the same way.

I see :)

> 
> > However, it seems very complex and i'm wary of it hiding other problems. If we're going to do it this way, it needs to be tested to destruction.
> 
> I totally agree, hence the solution we discussed about the `TwitterClient`.
> There are also limitations with what this also give us. For example it becomes very complex to handle ThreadContext within an AsyncContext and the other way around.

I don't understand what happens to `ThreadLocal` values if you thread hop during an `async` operation. I'd *assume* the `ThreadLocal` value could change as you'd changed thread, but don't know if you are somehow protected against this.  
Do you know what happens?

With the `ThreadLocal` value being used first, and only then falling back to `AsyncLocal`, if there is any risk of the `ThreadLocal` value changing due to the thread changing, we need to understand it :)

> 
> > I'd be tempted to change this to be purely async local and document how to go about using it. I can't think of a reason for users to need to update their settings asynchronously, so if we change the above example to:
> 
> This could make sense, but I thought it would be confusing for users to have 2 different approach to a same problem as we would have Thread created by async/await not behaving the same way as Threads created by the user.

Things like `Task.Run` play quite nicely with `AsyncLocal`. Is the issue with `Thread.Start`? I can't remember if this copies the execution context?  

If we went down this route, we'd get rid of having to prepare the execution context, which would make this problem much simpler.